File: Sample_CIS_Client_Apps/java/CIS 2.0/src/main/java/ApiClient.java
import com.fasterxml.jackson.databind.ObjectMapper;

import javax.net.ssl.SSLContext;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.KeyManagementException;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.util.List;
import java.util.UUID;

import org.apache.hc.client5.http.classic.methods.HttpGet;
import org.apache.hc.client5.http.classic.methods.HttpPost;
import org.apache.hc.client5.http.classic.methods.HttpPut;
import org.apache.hc.client5.http.impl.classic.CloseableHttpClient;
import org.apache.hc.client5.http.impl.classic.CloseableHttpResponse;
import org.apache.hc.client5.http.impl.classic.HttpClients;
import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManagerBuilder;
import org.apache.hc.client5.http.ssl.DefaultClientTlsStrategy;
import org.apache.hc.client5.http.ssl.NoopHostnameVerifier;
import org.apache.hc.client5.http.entity.mime.MultipartEntityBuilder;
import org.apache.hc.client5.http.ssl.TrustAllStrategy;
import org.apache.hc.core5.http.ContentType;
import org.apache.hc.core5.http.HttpEntity;
import org.apache.hc.core5.http.ParseException;
import org.apache.hc.core5.http.io.entity.EntityUtils;
import org.apache.hc.core5.http.Header;
import org.apache.hc.core5.ssl.SSLContextBuilder;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;


public class ApiClient {
    private final CloseableHttpClient httpClient;
    private final String apiKeyHeader;
    private final String apiKey;
    private final String basePath;
    private final ObjectMapper objectMapper = new ObjectMapper().registerModule(new JavaTimeModule());

    public ApiClient(String baseUrl, String apiKeyHeader, String apiKey, boolean trustCerts) {
        this.apiKeyHeader = apiKeyHeader;
        this.apiKey = apiKey;
        this.basePath = baseUrl.endsWith("/") ? baseUrl + "api/v2/ClientIntegration/" : baseUrl + "/api/v2/ClientIntegration/";

        var connectionManagerBuilder = PoolingHttpClientConnectionManagerBuilder.create();

        if (trustCerts) {
            // Bypass SSL certificate validation and hostname verification
            SSLContext sslContext;
            try {
                sslContext = SSLContextBuilder.create()
                        .loadTrustMaterial(TrustAllStrategy.INSTANCE)
                        .build();
            } catch (NoSuchAlgorithmException | KeyStoreException | KeyManagementException e) {
                throw new RuntimeException("Failed to create SSL context", e);
            }

            var tlsStrategy = new DefaultClientTlsStrategy(
                    sslContext,
                    NoopHostnameVerifier.INSTANCE);

            connectionManagerBuilder.setTlsSocketStrategy(tlsStrategy);
        }

        var connectionManager = connectionManagerBuilder.build();

        this.httpClient = HttpClients.custom()
                .setConnectionManager(connectionManager)
                .build();
    }

    public EnvironmentResponse getEnvironment() throws IOException, ParseException {
        HttpGet get = new HttpGet(basePath + "Environment");
        get.addHeader(apiKeyHeader, apiKey);

        try (CloseableHttpResponse response = httpClient.execute(get)) {
            if (response.getCode() < 200 || response.getCode() > 204) {
                throw new IOException("Failed : HTTP error code : " + response.getCode());
            }
            String body = EntityUtils.toString(response.getEntity());
            return objectMapper.readValue(body, EnvironmentResponse.class);
        }
    }

    public UUID submit(UUID repositoryId, List<String> inputFilePaths) throws IOException, ParseException {
        HttpPost post = new HttpPost(basePath + "Submit");
        post.addHeader(apiKeyHeader, apiKey);

        MultipartEntityBuilder builder = MultipartEntityBuilder.create();
        builder.addTextBody("RepositoryId", repositoryId.toString());

        // Use InputStreams so the multipart entity can stream the file content
        // Keep references to streams so we can close them after the request completes
        List<java.io.InputStream> streams = new java.util.ArrayList<>();
        try {
            for (int i = 0; i < inputFilePaths.size(); i++) {
                Path path = Paths.get(inputFilePaths.get(i));
                String fileName = path.getFileName().toString();
                java.io.InputStream is = Files.newInputStream(path);
                streams.add(is);
                builder.addBinaryBody("InputFiles[" + i + "].InputFile", is, ContentType.APPLICATION_OCTET_STREAM, fileName);
                builder.addTextBody("InputFiles[" + i + "].FileMetadata[0].Name", "Java Sample App Submission");
                builder.addTextBody("InputFiles[" + i + "].FileMetadata[0].Value", "Test file uploaded via Java sample app");

                // or if you want to add more metadata, repeat the above two lines with different indices when 1:1 size ratio of lists
                /*
                List<List<Metadata>> m = new ArrayList<>();
                m.add(new ArrayList<Metadata>(new Metadata("Name1", "Value1")));

                for (int j = 0; j < m.get(i).size(); j++) {
                    builder.addTextBody("InputFiles[" + i + "].FileMetadata[" + j + "].Name", m.get(i).get(j).getName());
                    builder.addTextBody("InputFiles[" + i + "].FileMetadata[" + j + "].Value", m.get(i).get(j).getValue());
                }
                */
            }

            HttpEntity entity = builder.build();
            post.setEntity(entity);

            try (CloseableHttpResponse response = httpClient.execute(post)) {
                if (response.getCode() < 200 || response.getCode() > 204) {
                    throw new IOException("Failed : HTTP error code : " + response.getCode());
                }
                String body = EntityUtils.toString(response.getEntity());
                return objectMapper.readValue(body, UUID.class);
            }
        } finally {
            // Ensure all streams are closed to avoid resource leaks
            for (java.io.InputStream s : streams) {
                try {
                    if (s != null) s.close();
                } catch (IOException ignored) {
                }
            }
        }
    }

    public JobStatusResponse getStatus(UUID jobId) throws IOException, ParseException {
        HttpGet get = new HttpGet(basePath + "Status/" + jobId);
        get.addHeader(apiKeyHeader, apiKey);

        try (CloseableHttpResponse response = httpClient.execute(get)) {
            if (response.getCode() < 200 || response.getCode() > 204) {
                throw new IOException("Failed : HTTP error code : " + response.getCode());
            }
            String body = EntityUtils.toString(response.getEntity());
            return objectMapper.readValue(body, JobStatusResponse.class);
        }
    }

    public void download(UUID jobId, String downloadDirectory) throws IOException {
        HttpGet get = new HttpGet(basePath + "Download/" + jobId);
        get.addHeader(apiKeyHeader, apiKey);

        try (CloseableHttpResponse response = httpClient.execute(get)) {
            if (response.getCode() < 200 || response.getCode() > 204) {
                throw new IOException("Failed : HTTP error code : " + response.getCode());
            }

            String fileName = jobId + ".unknown";
            Header cdHeader = response.getFirstHeader("Content-Disposition");
            String contentDisposition = cdHeader != null ? cdHeader.getValue() : null;
            if (contentDisposition != null) {
                String[] parts = contentDisposition.split(";");
                for (String part : parts) {
                    if (part.trim().startsWith("filename=")) {
                        fileName = part.trim().substring("filename=".length()).replace("\"", "");
                        break;
                    }
                }
            }

            Path filePath = Paths.get(downloadDirectory, fileName);
            try (InputStream is = response.getEntity().getContent()) {
                Files.copy(is, filePath);
            }
        }
    }

    public void release(UUID jobId) throws IOException {
        HttpPut put = new HttpPut(basePath + "Release/" + jobId);
        put.addHeader(apiKeyHeader, apiKey);

        try (CloseableHttpResponse response = httpClient.execute(put)) {
            if (response.getCode() < 200 || response.getCode() > 204) {
                throw new IOException("Failed : HTTP error code : " + response.getCode());
            }
        }
    }
}

File: Sample_CIS_Client_Apps/csharp/CIS 2.0/ApiClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Threading.Tasks;


namespace CIS_2_0
{
    public class ApiClient
    {
        private readonly HttpClient _httpClient;
        private readonly string _apiKeyHeader;
        private readonly string _apiKey;
        private readonly string _basePath;

        public ApiClient(string baseUrl, string apiKeyHeader, string apiKey, bool skipCertificateVerification = false)
        {
            if (skipCertificateVerification)
            {
                var handler = new HttpClientHandler
                {
                    ServerCertificateCustomValidationCallback = (message, cert, chain, errors) => true // Bypasses all certificate validation
                };

                _httpClient = new HttpClient(handler) { };
            }
            else
            {
                _httpClient = new HttpClient() { };
            }

            _apiKeyHeader = apiKeyHeader;
            _apiKey = apiKey;
            _basePath = baseUrl.EndsWith('/') ? $"{baseUrl}api/v2/ClientIntegration/" : $"{baseUrl}/api/v2/ClientIntegration/";
        }


        /// <summary>
        /// Adds the API key header to the HTTP request.
        /// </summary>
        /// <param name="request"></param>
        private void AddApiKeyHeader(HttpRequestMessage request)
        {
            request.Headers.Add(_apiKeyHeader, _apiKey);
        }


        /// <summary>
        /// Gets environment information for API key including repositories that are available to submit to based on your allowed
        /// workspaces and global variables.
        /// </summary>
        /// <returns></returns>
        public async Task<EnvironmentResponse> GetEnvironmentAsync()
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, $"{_basePath}Environment");
            AddApiKeyHeader(request);
            var response = await _httpClient.SendAsync(request);
            response.EnsureSuccessStatusCode();
            return await response.Content.ReadFromJsonAsync<EnvironmentResponse>();
        }


        /// <summary>
        /// Submits a job with the specified repository ID and input files.
        /// </summary>
        /// <param name="repositoryId"></param>
        /// <param name="inputFilePaths"></param>
        /// <returns></returns>
        public async Task<Guid> SubmitAsync(Guid repositoryId, List<string> inputFilePaths)
        {
            using var content = new MultipartFormDataContent();
            content.Add(new StringContent(repositoryId.ToString()), "RepositoryId");

            // Keep track of FileStreams to dispose after the request
            var fileStreams = new List<FileStream>();

            try
            {
                for (int i = 0; i < inputFilePaths.Count; i++)
                {
                    var filePath = inputFilePaths[i];
                    var fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read, FileShare.Read);
                    fileStreams.Add(fileStream); // Track for later disposal

                    var fileContent = new StreamContent(fileStream);
                    fileContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
                    content.Add(fileContent, $"InputFiles[{i}].InputFile", Path.GetFileName(filePath));

                    
                    content.Add(new StringContent("C# Sample App Submission"), $"InputFiles[{i}].FileMetadata[0].Name");
                    content.Add(new StringContent("Test file uploaded via C# sample app"), $"InputFiles[{i}].FileMetadata[0].Value");

                    //below is an example on how to add per file metadata entries via 1:1 size lists (metadata and files)
                    /*
                     * metadata is not used in this sample, but here's how you would add multiple metadata entries per file
                     for(int j = 0; j < metadata[i].Count; j++) {
                        content.Add(new StringContent(content.Add(new StringContent(metadata[i][j].Name), $"InputFiles[{i}].FileMetadata[{j}].Name");
                        content.Add(new StringContent(content.Add(new StringContent(metadata[i][j].Value), $"InputFiles[{i}].FileMetadata[{j}].Value");
                     }

                     */
                }

                using var request = new HttpRequestMessage(HttpMethod.Post, $"{_basePath}Submit");
                AddApiKeyHeader(request);
                request.Content = content;
                var response = await _httpClient.SendAsync(request);
                response.EnsureSuccessStatusCode();
                return await response.Content.ReadFromJsonAsync<Guid>();
            }
            finally
            {
                // Dispose all FileStreams after the request is complete
                foreach (var fileStream in fileStreams)
                {
                    fileStream.Dispose();
                }
            }
        }


        /// <summary>
        /// Gets the status of a submitted job by its Job ID.
        /// </summary>
        /// <param name="jobId"></param>
        /// <returns></returns>
        public async Task<JobStatusResponse> GetStatusAsync(Guid jobId)
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, $"{_basePath}Status/{jobId}");
            AddApiKeyHeader(request);
            var response = await _httpClient.SendAsync(request);
            response.EnsureSuccessStatusCode();
            return await response.Content.ReadFromJsonAsync<JobStatusResponse>();
        }


        /// <summary>
        /// Downloads the result files of a completed job to the specified download directory.
        /// </summary>
        /// <param name="jobId"></param>
        /// <param name="downloadDirectory"></param>
        /// <returns></returns>
        public async Task DownloadAsync(Guid jobId, string downloadDirectory)
        {
            using var request = new HttpRequestMessage(HttpMethod.Get, $"{_basePath}Download/{jobId}");
            AddApiKeyHeader(request);
            var response = await _httpClient.SendAsync(request);
            response.EnsureSuccessStatusCode();

            var contentDisposition = response.Content.Headers.ContentDisposition;
            var fileName = contentDisposition?.FileName.Trim('\"') ?? $"{jobId}.unknown";


            var filePath = Path.Combine(downloadDirectory, fileName);
            using var fileStream = new FileStream(filePath, FileMode.Create, FileAccess.Write, FileShare.None);
            await response.Content.CopyToAsync(fileStream);
        }


        /// <summary>
        /// Releases the job. Once released, job data is no longer available.
        /// </summary>
        /// <param name="jobId"></param>
        /// <returns></returns>
        public async Task ReleaseAsync(Guid jobId)
        {
            using var request = new HttpRequestMessage(HttpMethod.Put, $"{_basePath}Release/{jobId}");
            AddApiKeyHeader(request);
            var response = await _httpClient.SendAsync(request);
            response.EnsureSuccessStatusCode();
        }
    }
}
